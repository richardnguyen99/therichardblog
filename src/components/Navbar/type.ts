import * as React from "react";

export interface NavbarProps {
  /**
   * @description Fires a callback when the navbar has shrank
   */
  onShrink?: () => void;

  /**
   * @description Controls whether or not the navbar is shrinkable on scroll
   *
   * @default false
   */
  shrinkable?: boolean;

  /**
   * @description Controls whehter or not the navbar is fixed on the screen
   *
   * @default false
   */
  fixed?: boolean;

  [key: string]: any; // For spreading props
}

export interface NavbarItemProps {
  /**
   * @description Controls whether or not the navbar item will take the whole space left in the navbar
   *
   * @default false
   */
  full?: boolean;

  [key: string]: any;
}

export interface NavbarLinkProps {
  /**
   * @description Controls which type of HTML element the component should be rendered into.
   *
   * @see https://github.com/emotion-js/emotion/issues/1137#issuecomment-615571356 for more detail why
   *
   * @default a (as anchor tag)
   */
  as?: React.ElementType | keyof JSX.IntrinsicElements;

  /**
   * @description Controls the target url that the component should navigate to.
   *
   * @default "" (use to return to the main page)
   */
  href?: string;

  /**
   * @description Controls the internaltarget url the component should navigate to. `href` will override this property
   *
   * @default ""
   */
  to?: string;

  [key: string]: any; // For spreading props
}
