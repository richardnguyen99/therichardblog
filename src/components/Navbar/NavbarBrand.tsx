/**
 * NavbarBrand to contain the brand and its configuration
 *
 * @author Richard Nguyen <richard.ng0616@gmail.com>
 */

import * as React from "react";
import styled from "styled-components";
import { Link } from "gatsby";

const StyledNavbarBrand = styled(Link)`
  text-decoration: none;
  color: #fff;

  font-family: "Abril Fatface", Arial, Helvetica, sans-serif;
  font-size: var(--fs-h4);
`;

const NavbarBrand: React.FC = () => {
  return <StyledNavbarBrand to="/">Richar&apos;s blog</StyledNavbarBrand>;
};

export default NavbarBrand;
