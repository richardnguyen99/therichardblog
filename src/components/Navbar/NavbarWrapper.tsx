/**
 * NavbarWrapper to hold position of Navbar component
 *
 * @author Richard Nguyen <richard.ng0616@gmail.com>
 */

import * as React from "react";
import styled from "styled-components";

const StyledNavbarTopLevelWrapper = styled.nav`
  width: 100%;
  position: fixed;
`;

const StyledNavbarNavigationContainer = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-between;
  padding-top: 1rem;
  padding-bottom: 1rem;
  font-size: var(--fs-small);

  max-width: 1280px;
  margin-right: auto;
  margin-left: auto;
  padding-right: 1rem;
  padding-left: 1rem;
`;

const NavbarWrapper: React.FC = ({ children }) => {
  return (
    <StyledNavbarTopLevelWrapper>
      <StyledNavbarNavigationContainer>{children}</StyledNavbarNavigationContainer>
    </StyledNavbarTopLevelWrapper>
  );
};

export default NavbarWrapper;
