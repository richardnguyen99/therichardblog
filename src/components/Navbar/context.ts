/**
 * Context API for Navbar component
 *
 * @author Richard Nguyen <richard.ng0616@gmail.com>
 */
import * as React from "react";

import { NavbarContextValue } from "./type";

const NavbarContext = React.createContext<NavbarContextValue>({});
NavbarContext.displayName = "NavbarContext";

export default NavbarContext;
