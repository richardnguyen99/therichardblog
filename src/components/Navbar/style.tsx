/**
 * Styled components for Navbar
 *
 * This file should only contains style definitions of the navbar components, do not handle logic here
 *
 * @author Richard Nguyen <richard.ng0616@gmail.com>
 */
import styled from "styled-components";

export const StyledNavbar = styled.nav`
  display: flex;
  padding: 1rem;
  font-size: 14px;
  line-height: 1.5;
  color: rgba(255, 255, 255, 1); // TODO: Implement color text for light theme
  background-color: #18181b; // TODO: Implement color background for light theme
  align-items: center;
  flex-wrap: nowrap;
`;

export const StyledNavbarItem = styled.div<{ full?: boolean }>`
  display: flex;
  margin-right: 1rem;
  align-self: center;
  flex-wrap: nowrap;

  &:last-child {
    margin-right: 0;
  }

  ${(props) => props.full && "flex: auto;"}
`;

export const StyledNavbarLink = styled.a`
  display: inline-flex;
  align-items: center;
  font-weight: 600;
  color: rgba(255, 255, 255, 1); // TODO: Implement color text for light theme
  white-space: nowrap;
  cursor: pointer;
  text-decoration: none;

  padding-top: 8px;
  padding-bottom: 8px;
  margin-right: 32px;

  &:hover,
  &:focus {
    color: rgba(255, 255, 255, 0.8); // TODO: implment color text hover for light theme
  }
`;

export const StyledNavbarBrand = styled(StyledNavbarLink)`
  display: flex;
  align-items: center;

  svg,
  .icon {
    margin-right: 8px;
    fill: currentColor;
    overflow: visible;
  }
`;
