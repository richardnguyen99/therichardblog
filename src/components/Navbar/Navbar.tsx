/**
 * Navbar to represent the main navbar header component in app
 *
 * @author Richard Nguyen <richard.ng0616@gmail.com>
 */

import * as React from "react";

import NavbarBrand from "./NavbarBrand";
import NavbarWrapper from "./NavbarWrapper";

const Navbar: React.FC = () => {
  return (
    <NavbarWrapper>
      <NavbarBrand />
    </NavbarWrapper>
  );
};

export default Navbar;
