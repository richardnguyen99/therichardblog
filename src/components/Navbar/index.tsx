/**
 * Core navbar components
 *
 * @author Richard Nguyen <richard.ng0616@gmail.com>
 */
import * as React from "react";
import { Link } from "gatsby";

import { NavbarProps, NavbarLinkProps, NavbarItemProps } from "./type";
import { StyledNavbar, StyledNavbarBrand, StyledNavbarItem, StyledNavbarLink } from "./style";

/**
 * @description Navigates to other pages and sites
 */
const NavbarLink: React.FC<NavbarLinkProps> = React.forwardRef<
  HTMLAnchorElement | HTMLSpanElement,
  NavbarLinkProps
>(({ as, href = "", to = "", children, ...rest }, ref) => {
  const Component = as || (href ? "a" : to ? Link : "span");

  return (
    <StyledNavbarLink ref={ref} as={Component} href={href} to={!href ? to : undefined} {...rest}>
      {children}
    </StyledNavbarLink>
  );
});
NavbarLink.displayName = "NavbarLink";

/**
 * @description Controls the displayment of the brand
 */
const NavbarBrand: React.FC<NavbarLinkProps> = React.forwardRef<
  HTMLAnchorElement | HTMLSpanElement,
  NavbarLinkProps
>(({ as, href = "", to = "", children, ...rest }, ref) => {
  const Component = as || (href ? "a" : to ? Link : "span");

  return (
    <StyledNavbarBrand ref={ref} as={Component} href={href} to={!href ? to : undefined} {...rest}>
      {children}
    </StyledNavbarBrand>
  );
});
NavbarBrand.displayName = "NavbarBrand";

/**
 * @description Contains an individual or a group of components
 */
const NavbarItem: React.FC<NavbarItemProps> = React.forwardRef<HTMLDivElement, NavbarItemProps>(
  ({ full = false, children, ...rest }, ref) => {
    return (
      <StyledNavbarItem ref={ref} full={full} {...rest}>
        {children}
      </StyledNavbarItem>
    );
  }
);
NavbarItem.displayName = "NavbarItem";

/**
 * @description Displays the navbar component in the site
 */
const Navbar: React.FC<NavbarProps> = React.forwardRef<HTMLElement, NavbarProps>(
  ({ children, ...rest }, ref) => {
    return (
      <StyledNavbar ref={ref} {...rest}>
        {children}
      </StyledNavbar>
    );
  }
);
Navbar.displayName = "Navbar";

export default Object.assign(Navbar, { Item: NavbarItem, Link: NavbarLink, Brand: NavbarBrand });
