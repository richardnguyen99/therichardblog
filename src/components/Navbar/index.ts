/**
 * Global export for Navbar components
 *
 * @author Richard Nguyen <richard.ng0616@gmail.com>
 */

import Navbar from "./Navbar";
import NavbarBrand from "./NavbarBrand";
import NavbarWrapper from "./NavbarWrapper";

export default Object.assign(Navbar, {
  Wrapper: NavbarWrapper,
  Brand: NavbarBrand,
});
