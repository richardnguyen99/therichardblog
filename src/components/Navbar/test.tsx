/**
 * Core test file for navbar component
 *
 * @author Richard Nguyen <richard.ng0616@gmail.com>
 *
 * @jest-environment jsdom
 */

import * as React from "react";
import { render } from "@testing-library/react";

import Navbar from "./index";
import Icon from "../icons";

describe("<Navbar />", () => {
  it("should match snapshot", () => {
    const { container } = render(
      <Navbar>
        <Navbar.Brand to="/">Richard&apos;s blog</Navbar.Brand>
        <Navbar.Item full>
          <Navbar.Link to="/about">About</Navbar.Link>
          <Navbar.Link to="/post">About</Navbar.Link>
        </Navbar.Item>
        <Navbar.Item>
          <Navbar.Link href="https://github.com/richardnguyen99/therichardblog">Github</Navbar.Link>
        </Navbar.Item>
      </Navbar>
    );

    expect(container).toMatchSnapshot();
  });
});

describe("<Navbar.Item>", () => {
  it("should be a flex container", () => {
    const { container } = render(<Navbar.Item>Test</Navbar.Item>);

    expect(container.firstChild).toHaveStyle("display: flex");
  });

  it("should be a full flex container", () => {
    const { container } = render(<Navbar.Item full>Test</Navbar.Item>);

    expect(container.firstChild).toHaveStyle("display: flex; flex: auto");
  });
});

describe("<Navbar.Link>", () => {
  it("should not be a flex container", () => {
    const { container } = render(<Navbar.Link to="/">Link</Navbar.Link>);

    expect(container.firstChild).not.toHaveStyle("display: flex");
  });

  it("should render properly with link tag", () => {
    const { container, getByText, getByRole } = render(<Navbar.Link to="/">Link</Navbar.Link>);

    expect(getByText(/Link/i).closest("a")).toBeInTheDocument();
    expect(getByRole("link")).toBeInTheDocument();
    expect(container.firstChild.nodeName).toBe("A");
    expect(container.firstChild).toHaveAttribute("href", "/");
  });

  it("should render properly with span tag", () => {
    const { container, getByText } = render(<Navbar.Link>Link</Navbar.Link>);

    expect(getByText(/Link/i)).toBeInTheDocument();
    expect(container.firstChild.nodeName).toBe("SPAN");
  });
});

describe("<Navbar.Brand>", () => {
  it("should be a flex container", () => {
    const { container } = render(<Navbar.Brand>Test</Navbar.Brand>);

    expect(container.firstChild).toHaveStyle("display: flex");
  });
  it("should render properly with link tag", () => {
    const { container, getByText, getByRole } = render(<Navbar.Brand to="/">Link</Navbar.Brand>);

    expect(getByText(/Link/i).closest("a")).toBeInTheDocument();
    expect(getByRole("link")).toBeInTheDocument();
    expect(container.firstChild.nodeName).toBe("A");
    expect(container.firstChild).toHaveAttribute("href", "/");
  });

  it("should render properly with span tag", () => {
    const { container, getByText } = render(<Navbar.Brand>Link</Navbar.Brand>);

    expect(getByText(/Link/i)).toBeInTheDocument();
    expect(container.firstChild.nodeName).toBe("SPAN");
  });

  it("should render with icon child", () => {
    const { container, getByText } = render(
      <Navbar.Brand href="https://github.com/richardnguyen99/therichardblog">
        <Icon.Github />
        Github
      </Navbar.Brand>
    );

    expect(container.firstChild.firstChild.nodeName).toBe("svg");
    expect(getByText(/Github/i)).toBeInTheDocument();
    expect(container.firstChild.nodeName).toBe("A");
    expect(container.firstChild).toHaveAttribute(
      "href",
      "https://github.com/richardnguyen99/therichardblog"
    );
  });
});
