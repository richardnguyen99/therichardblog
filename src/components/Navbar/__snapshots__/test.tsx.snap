// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Navbar /> should match snapshot 1`] = `
<div>
  <nav
    class="sc-bdfBQB MKTIf"
  >
    <a
      class="sc-dlfnuX sc-hKgJUU iWCnRH eOFVTJ"
      href="/"
    >
      Richard's blog
    </a>
    <div
      class="sc-gsTEea evDERg"
    >
      <a
        class="sc-dlfnuX iWCnRH"
        href="/about"
      >
        About
      </a>
      <a
        class="sc-dlfnuX iWCnRH"
        href="/post"
      >
        About
      </a>
    </div>
    <div
      class="sc-gsTEea hhKWjN"
    >
      <a
        class="sc-dlfnuX iWCnRH"
        href="https://github.com/richardnguyen99/therichardblog"
      >
        Github
      </a>
    </div>
  </nav>
</div>
`;
